Fix plugin config per environment doesn't apply

iicdii opened the pull request:
Signed-off-by: harimkims <harimkims@gmail.com>

<!--
Hello 👋 Thank you for submitting a pull request.

To help us merge your PR, make sure to follow the instructions below:

- Create or update the documentation. (Should be made against the documentation branch)
- Create or update the tests.
- Refer to the issue you are closing in the PR description - fix #issue
- Specify if the PR is in WIP (work in progress) state or ready to be merged

Please ensure you read through the Contributing Guide: https://github.com/strapi/strapi/blob/master/CONTRIBUTING.md
-->

### What does it do?

I added logic to find `config/env/<process.env.NODE_ENV>/plugins.js` first when it loads plugin config.
**As-is**: `config/plugins.js` -> default plugin config
**To-be**: `config/env/${process.env.NODE_ENV}/plugins.js` -> `config/plugins.js` -> default plugin config

### Why is it needed?

To support [Environment configurations](https://docs.strapi.io/developer-docs/latest/setup-deployment-guides/configurations/optional/environment.html#environment-configurations)

### How to test it?

1. Create `config/env/development/plugins.js`
2. Copy and taste config below
```js
module.exports = () => ({
  upload: {
    enabled: true,
    config: {
      provider: 'local',
      providerOptions: {
        sizeLimit: 3000000,
      },
    },
  },
});
```
3. Start interactive CLI `yarn strapi console`
4. Try `strapi.config.get('plugin.upload', {})` and see difference
5. Create `config/plugins.js`
6. Repeat Step 2~4 but change `sizeLimit` to 2000000
7. Delete `config/plugins.js`
8. Repeat Step 3~4 (default `sizeLimit` is 1000000 though)

### Related issue(s)/PR(s)

Closes #11774

alexandrebodin reviewed (changes requested):
Ideally, we would like this to work the same way as other configurations, 

Loading the main plugins.js file and merging the env/x/plugins.js file with it instead of switching. This would make the behavior more consistent :)

iicdii commented:
I applied your feedback. What do you think?

alexandrebodin commented on packages/core/strapi/lib/core/loaders/plugins/index.js:
```diff
@@ -65,11 +65,29 @@ const formatContentTypes = plugins => {
+  const globalUserConfigPath = join(strapi.dirs.config, 'plugins.js');
+  const currentEnvUserConfigPath = join(
+    strapi.dirs.config,
+    `env/${process.env.NODE_ENV}/plugins.js`
```
to simplify you can do 
```js
join(strapi.dirs.config, 'env', process.env.NODE_ENV, 'plugins.js')
```

alexandrebodin reviewed (commented):
Just a simplification in the code but that looks quite good. I'll test it

alexandrebodin reviewed (commented):
Something that will not be working is disabling a plugin in an env. The getEnablePlugin code loads the config again. I think this deserves some refactoring :)

iicdii commented:
I changed 2 parts.
1. in `get-enabled-plugins.js` function, DeclaredPlugins didn't look for config exist in `config/env/${env}/plugins.js`.. I fixed it. Now it cares about user's config by environment.
2. I found that the `getEnablePlugin` code enables every plugin whether user configured it `true` or `false` or whatever. I fixed it so now user's `enabled: false` flag is applied to config. The problem was Lodash `defaultsDeep` function ignores the property if it has `enabled` key already. I kept `defaultsDeep` function, but I applied user's enabled flag after that.

alexandrebodin commented:
> I reproduced the problem you said. I changed 2 parts.
> 
> 1. I fixed `declaredPlugins` in `get-enabled-plugins.js` function didn't look for config exist in `config/env/${env}/plugins.js`. Now it cares about user's config by environment.
> 2. I found that the `getEnablePlugin` code enables every plugin whether user configured it `true` or `false` or whatever. I fixed it so now user's `enabled: false` flag is applied to config. The problem was Lodash `defaultsDeep` function ignores the property if it has `enabled` key already. I kept `defaultsDeep` function and I applied user's enabled flag after that.

I never had an issue with disable a plugin I'm not sure this is accurate 🤔

iicdii commented:
You're right.

> I found that the getEnablePlugin code enables every plugin whether user configured it true or false or whatever. I fixed it so now user's enabled: false flag is applied to config. The problem was Lodash defaultsDeep function ignores the property if it has enabled key already. I kept defaultsDeep function and I applied user's enabled flag after that.

It was wrong. I reverted this fix to the original code. However, the first modification - `get-enabled-plugins.js` function didn't look for config exist in `config/env/${env}/plugins.js`  - is still valid.

alexandrebodin commented on packages/core/strapi/lib/core/loaders/plugins/get-enabled-plugins.js:
```diff
@@ -106,6 +103,7 @@ const getEnabledPlugins = async strapi => {
 
   const enabledPlugins = pipe(
     defaultsDeep(declaredPlugins),
+    pickBy((pValue, pName) => _.get(declaredPlugins, [pName, 'enabled'], true)),
```
is this necessary? the last pickBy in the list should be doing the job

iicdii replied:
No It's not, I just pushed commit

alexandrebodin reviewed (approved):
This works like a charm. Thanks you for this great contribution !
